api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto

```diff
@@ -10,11 +12,31 @@ option java_outer_classname = "OnDemandProto";
 option java_multiple_files = true;
 option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
-// [#protodoc-title: OnDemand]
-// IP tagging :ref:`configuration overview <config_http_filters_on_demand>`.
+// [#protodoc-title: On Demand Discovery]
+// On Demand Discovery :ref:`configuration overview <config_http_filters_on_demand>`.
 // [#extension: envoy.filters.http.on_demand]
 
+// On Demand Discovery filter config.
 message OnDemand {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.http.on_demand.v2.OnDemand";
+
+  // An optional configuration for on-demand cluster discovery
+  // service. If not specified, the on-demand cluster discovery will
+  // be disabled. When it's specified, the filter will pause a request
+  // to an unknown cluster and will begin a cluster discovery
+  // process. When the discovery is finished (successfully or not),
+  // the request will be resumed for further processing.
+  config.core.v3.ConfigSource odcds_config = 1;
+}
+
+// Per-route configuration for On Demand Discovery.
+message PerRouteConfig {
```

Comment:
We need both the ability to disable this on a per-route basis and the ability to handle xdstp names for which no `ConfigSource` needs to be specified.  In https://github.com/envoyproxy/envoy/pull/15523/files/b742a88722c205b9fa1d0b694ff010c3d994e7b2#r596505958, @htuch proposed that we use an empty `PerRouteConfig` message for xdstp names; here, you are proposing that we use the same thing to indicate disabling the feature.  I don't think this can mean both things.

Reply:
Technically we could have an empty message with an empty `ConfigSource`, but I think this gets confusing. Probably an explicit disable flag would be the way to go.